### PR TITLE
EVP: Don't find standard EVP_PKEY_METHODs automatically

### LINKED ISSUE #11424: Falling back to legacy when we shouldn't
In some internal tests I just saw a TLS connection successfully perform an X25519 operation in a library context which only had the FIPS provider loaded - not the default provider. This is unexpected because X25519 does not exist in the FIPS provider (yet).

The reason seems to be because in our key exchange code we attempt to fetch the algorithm from the available providers. However if that does not work we fallback to legacy and see if the legacy path can help us. And of course in this case it can.

This is undesirable because if you only have the FIPS provider loaded you only want FIPS crypto being performed. Anything else should fail.

We should only really fallback to legacy if we know that an engine can service the request.

My guess is this same issue applies to other operations besides key exchange.

### PR DESCRIPTION
**Pull request by levitte**
EVP_PKEY_meth_find() got called automatically any time a new
EVP_PKEY_CTX allocator was called with some sort of key type data.
Since we have now moved all our standard algorithms to our providers,
this is no longer necessary.

We do retain looking up EVP_PKEY_METHODs that are added by the calling
application.

Fixes #11424

### CODE REVIEW

**Comment by beldmit** [comment]
No, not that simple, unfortunately.

**Comment by levitte** [comment]
> No, not that simple, unfortunately.

The fact that it uncovers bugs (which is what it does) is, well, unfortunate, but I'll maintain that the solution itself is simple.

**Comment by beldmit** [comment]
It also breaks gost-engine tests. So could you please add the extended tests to this PR if they are compatible with current CI?

**Comment by levitte** [comment]
I haven't the faintest idea why it would break the GOST engine...

**Comment by levitte** [comment]
I found a case when this breaks...  anytime someone calls `EVP_PKEY_assign()` or one of its wrapper macros.  I had forgotten about those, and those calls are also the time when we do fall back on EVP_PKEY_METHODs.

Closing this.

**Comment by levitte** [comment]
Being a little more awake now, I found that there's an oddity in `test/ectest.c` : `custom_params_test()`, where the keys with custom groups are unexportable (i.e., `evp_pkey_export_to_provider()`, which gets called by `EVP_PKEY_derive_init()`, fails, which means that `EVP_PKEY_derive_init()` jumps to the legacy section...  which needs a non-NULL `ctx->pmeth` to do its work.

The section of code that causes the export to fail is this:

https://github.com/openssl/openssl/blob/0c3eb2793b2a1fe35beeb90ba8f5cb2a0fdc3270/crypto/ec/ec_lib.c#L1698-L1714

The failure is caused by `encoding_flag == OPENSSL_EC_NAMED_CURVE`

...  and this is where I don't quite understand the ins and outs of the EC code, and would like some help figuring out this failure.  I thought that at this point, keys with custom groups could be handled by the provider, but maybe I misunderstood...  @romen?  @slontis?

**Comment by levitte** [comment]
You don't need this PR to get the export failure, it happens in master as well.  You only need a breakpoint here to see it in action:

https://github.com/openssl/openssl/blob/0c3eb2793b2a1fe35beeb90ba8f5cb2a0fdc3270/crypto/evp/exchange.c#L226

**Comment by beldmit** [comment]
GOST failures have gone.

**Comment by mattcaswell** [comment]
The ectest failure seems to have uncovered a latent bug in the EC library. It seems the "asn1_flag" setting was defaulting to "named curve", and you have to explicitly change it if you have set up an explicit curve. It seems to me that the default should be the other way around, and that we can change the "asn1_flag" when we know what the group name is.

The following patch fixes the test failure:

````
diff --git a/crypto/ec/ec_lib.c b/crypto/ec/ec_lib.c
index e1b92f7c33..508c96fef5 100644
--- a/crypto/ec/ec_lib.c
+++ b/crypto/ec/ec_lib.c
@@ -63,7 +63,7 @@ EC_GROUP *ec_group_new_ex(OSSL_LIB_CTX *libctx, const char *propq,
         if (ret->cofactor == NULL)
             goto err;
     }
-    ret->asn1_flag = OPENSSL_EC_NAMED_CURVE;
+    ret->asn1_flag = OPENSSL_EC_EXPLICIT_CURVE;
     ret->asn1_form = POINT_CONVERSION_UNCOMPRESSED;
     if (!meth->group_init(ret))
         goto err;
@@ -481,6 +481,7 @@ const BIGNUM *EC_GROUP_get0_cofactor(const EC_GROUP *group)
 void EC_GROUP_set_curve_name(EC_GROUP *group, int nid)
 {
     group->curve_name = nid;
+    group->asn1_flag = OPENSSL_EC_NAMED_CURVE;
 }
 
 int EC_GROUP_get_curve_name(const EC_GROUP *group)
````
Basically the test was creating an explicit params group, but it was still marked as using named group encoding. This was failing when it was being exported to the provider because the provider didn't have a named group that matched the explicit params.

An alternative fix would be to change the test to explicitly call `EC_GROUP_set_asn1_flag()` on the constructed group (`altgroup`).

**Comment by levitte** [comment]
I was looking at `asn1_flag` with quite a bit of suspicion, but didn't dare look too deeply.  Thanks for uncovering/confirming!  I'll apply your patch.

**Comment by levitte** [comment]
Ping!

File: crypto/ec/ec_lib.c [hunk 1]
```diff
@@ -481,6 +481,7 @@ const BIGNUM *EC_GROUP_get0_cofactor(const EC_GROUP *group)
 void EC_GROUP_set_curve_name(EC_GROUP *group, int nid)
 {
     group->curve_name = nid;
+    group->asn1_flag = OPENSSL_EC_NAMED_CURVE;
```
Comment (slontis) [inline crypto/ec/ec_lib.c:484]: Maybe do a check for undef?
````
group->asn1_flag = (nid != NID_undef) ? : ;
````

File: crypto/evp/pmeth_lib.c [hunk 2]
```diff
@@ -88,22 +88,34 @@ static int pmeth_cmp(const EVP_PKEY_METHOD *const *a,
...
 
-const EVP_PKEY_METHOD *EVP_PKEY_meth_find(int type)
+static const EVP_PKEY_METHOD *evp_pkey_meth_find_added_by_application(int type)
 {
-    pmeth_fn *ret;
     EVP_PKEY_METHOD tmp;
-    const EVP_PKEY_METHOD *t = &tmp;
 
     tmp.pkey_id = type;
     if (app_pkey_methods) {
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c:96]: != NULL

File: crypto/evp/pmeth_lib.c [hunk 3]
```diff
@@ -88,22 +88,34 @@ static int pmeth_cmp(const EVP_PKEY_METHOD *const *a,
...
 }
 
-const EVP_PKEY_METHOD *EVP_PKEY_meth_find(int type)
+static const EVP_PKEY_METHOD *evp_pkey_meth_find_added_by_application(int type)
 {
-    pmeth_fn *ret;
     EVP_PKEY_METHOD tmp;
-    const EVP_PKEY_METHOD *t = &tmp;
 
     tmp.pkey_id = type;
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c:95]: move inside the if block..

File: crypto/evp/pmeth_lib.c [hunk 4]
```diff
@@ -1665,8 +1677,33 @@ static int evp_pkey_ctx_store_cached_data(EVP_PKEY_CTX *ctx,
...
-        || ((optype != -1) && !(ctx->operation & optype))) {
+    if (keytype != -1) {
+        switch (evp_pkey_ctx_state(ctx)) {
+        case EVP_PKEY_STATE_PROVIDER:
+            if (ctx->keymgmt == NULL) {
+                ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+                return -2;
+            }
+            if (!EVP_KEYMGMT_is_a(ctx->keymgmt,
+                                  evp_pkey_type2name(keytype))) {
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c:1688]: For keytype = EVP_PKEY_DHX the name used would be "X9.42 DH"
Does this alias work with EVP_KEYMGMT_is_a() for "DHX"?

Comment (t8m) [reply crypto/evp/pmeth_lib.c:1688]: IMO yes. Some tests and some functionality already depend on this.

Comment (levitte) [reply crypto/evp/pmeth_lib.c:1688]: https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/providers/defltprov.c#L412-L413
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/providers/fips/fipsprov.c#L398-L399

File: crypto/ec/ec_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/ec/ec_lib.c:484]: Done

File: crypto/evp/pmeth_lib.c (hunk 2 above)
Comment (levitte) [reply crypto/evp/pmeth_lib.c:96]: Fixed

File: crypto/evp/pmeth_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/evp/pmeth_lib.c:95]: Yup

**Comment by levitte** [comment]
Ping!

**Comment by beldmit** [comment]
Will test tomorrow

**Comment by beldmit** [comment]
GOST tests are OK.

**Comment by levitte** [comment]
> GOST tests are OK.

Cool.  So, final review?

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by levitte** [comment]
Merged

60488d2434c5be15dc14e1fa2a8733f076d9ccf4 EVP: Don't find standard EVP_PKEY_METHODs automatically
977e95b912138d02bae86d829a990d81c2bbcca0 EVP: Fix evp_pkey_ctx_store_cached_data() to handle provider backed EVP_PKEY_CTX
9db6af922c48c5cab5398ef9f37e425e382f9440 EC: Reverse the default asn1_flag in a new EC_GROUP
